piper/mousemap.py

```diff
@@ -0,0 +1,390 @@
+
+        for child in self._children:
+            if not child.widget.get_visible():
+                continue
+            svg_geom = self._get_svg_sub_geometry(child.svg_leader)[1]
+            nat_size = child.widget.get_preferred_size()[1]
+            if child.is_left:
+                child_allocation.x = x + svg_geom.x - self.spacing - nat_size.width
+            else:
+                child_allocation.x = x + svg_geom.x + self.spacing
+            child_allocation.y = y + svg_geom.y + 0.5 * svg_geom.height - 0.5 * nat_size.height
+            child_allocation.width = nat_size.width
+            child_allocation.height = nat_size.height
+            if not child.widget.get_has_window():
+                child_allocation.x += allocation.x
+                child_allocation.y += allocation.y
+            child.widget.size_allocate(child_allocation)
+
+    def do_draw(self, cr):
+        """Draws the container to the given Cairo context. The top left corner
+        of the widget will be drawn to the currently set origin point of the
+        context. The container needs to propagate the draw signal to its
+        children.
+
+        @param cr The Cairo context to draw into, as cairo.Context
+        """
+        cr.save()
+        x, y = self._translate_to_origin()
+        cr.translate(x, y)
+        self._draw_device(cr)
+        cr.restore()
+        for child in self._children:
+            self.propagate_draw(child.widget, cr)
+
+    def do_get_property(self, prop):
+        """Gets a property value.
+
+        @param prop The property to get, as GObject.ParamSpec
+        """
+        if prop.name == "spacing":
+            return self.spacing
+        else:
+            raise AttributeError("Unknown property %s" % prop.name)
+
+    def _on_enter(self, widget, event, child):
+        # Highlights the element in the SVG to which the given widget belongs.
+        self._highlight_element = child.svg_id
+        self._redraw_svg_element(child.svg_id)
+
+    def _on_leave(self, widget, event):
+        # Restores the device SVG to its original state.
+        old_highlight = self._highlight_element
+        self._highlight_element = None
+        self._redraw_svg_element(old_highlight)
+
+    def _xpath_has_style(self, svg_id, style):
+        # Checks if the SVG element with the given identifier has the given
+        # style attribute set.
+        namespaces = {
+            'sodipodi': 'http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd',
+            'cc': 'http://web.resource.org/cc/',
+            'svg': 'http://www.w3.org/2000/svg',
+            'dc': 'http://purl.org/dc/elements/1.1/',
+            'xlink': 'http://www.w3.org/1999/xlink',
+            'rdf': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
+            'inkscape': 'http://www.inkscape.org/namespaces/inkscape'
+        }
+        query = "//svg:rect[@id=\"{}\"][contains(@style, \"{}\")]".\
+                format(svg_id, style)
```

Comment:
screw the 78 char limit here and make this one line please, much more obvious that way